src/lib/loaders/batchLoader.ts

```
@@ -59,39 +82,89 @@ interface BatchLoaderArgs {
+          }
+        })
+      )
+      const normalizedData = data.map(
+        datum => (Array.isArray(datum) ? datum.reverse() : [datum])
+      )
+      const results = idWithParamsList.map(params_2 => {
+        const paramGroup = serializeParams(params_2)
+        const groupIndex = paramGroups.indexOf(paramGroup)
+        return normalizedData[groupIndex].pop()
       })
-      return flatten(normalizedResults)
-    })
-  })
+      return results
+    },
+    {
+      maxBatchSize: 20,
+      cacheKeyFn,
+    }
+  )
+
+  type Key = string | { id: string[] }
+
+  /**
+   * This is an abstraction around the data loader to keep api parity with the
+   * existing gravityLoader api
+   */
+  return (key: Key) => {
+    /**
+     * This section covers the case when an endpoint is being requested that supports
+     * parameters. An example would the be `sales` endpoint which has filters like
+     * `live` or `is_auction`.
+     *
+     * The assumption here is that things are being requested from a list endpoint
+     * so the results are always formatted into an array
+     */
+    if (typeof key === "object" && key !== null) {
+      if (key.id.length === 1) {
+        return dl
+          .load({ ...key, id: key.id[0] })
+          .then(
+            result => (result === undefined || result === null ? [] : [result])
+          )
+      }
 
-  return key => dl.load(key)
+      return dl
+        .loadMany(
+          key.id.map(id => ({
+            ...key,
+            id: id,
+          }))
+        )
+        .then(results => results.filter(result => result !== null))
+      /**
+       * This section covers the case when a single resource endpoint is being requested.
+       * Take the `sale` endpoint for example. `key` is expected to be a string of `id` or `slug`.
+       */
+    } else if (typeof key === "string") {
+      return dl.load({ id: key })
+    } else {
+      console.error("Requested an invalid key type for batchLoader")
```

Why not throw an error instead?

---

```suggestion
      throw new Error("Requested an invalid key type for batchLoader")
```

---

Like that?

---

Yup! I’d rather fail early when an API isn’t being used right, especially when there’s types that should prevent it.

If we were using FB’s way of doing [assertions](https://en.wikipedia.org/wiki/Assertion_(software_development)), which I’d really like, this would be classed as an `invariant`.